R/glmCommonFunctions.R

```diff
@@ -55,34 +55,70 @@
   # make formulas for the full model and the null model
   ff <- .createGLMFormula(options, nullModel = FALSE)
   nf <- .createGLMFormula(options, nullModel = TRUE)
-  # specify family and link
-  if (options$family == "bernoulli") {
-    family <- "binomial"
-  }
-  else if (options$family == "gamma") {
-    family <- "Gamma"
-  }
-  else if (options$family == "inverseGaussian") {
-    family <- "inverse.gaussian"
-  }
-  else {
-    family <- options$family
-  }
 
-  familyLink <- eval(call(family, link = options$link))
-  # compute full and null models
-  if (options$weights == "") {
-    fullModel <- stats::glm(ff, family = familyLink, data = dataset, weights = NULL)
-    nullModel <- stats::glm(nf, family = familyLink, data = dataset, weights = NULL)
-  } else {
+  if (options$family != "other") {
+    # specify family and link
+    if (options$family == "bernoulli") {
+      family <- "binomial"
+    }
+    else if (options$family == "gamma") {
+      family <- "Gamma"
+    }
+    else if (options$family == "inverseGaussian") {
+      family <- "inverse.gaussian"
+    }
+    else {
+      family <- options$family
+    }
+
+    familyLink <- eval(call(family, link = options$link))
+    # compute full and null models
     fullModel <- stats::glm(ff,
                             family = familyLink,
                             data = dataset,
-                            weights = get(options$weights))
+                            weights = eval(.getWeightVariable(options$weights)))
     nullModel <- stats::glm(nf,
                             family = familyLink,
                             data = dataset,
-                            weights = get(options$weights))
+                            weights = eval(.getWeightVariable(options$weights)))
```

Comment:
Why is this necessary? Could you just pass the vector of values from the dataset?

Reply:
Hi Simon, thanks for the quick review!

1. Yes the implementation has some rough edges. I haven't been able to load the module into JASP (see screenshot for error, any idea what happened?), so it wasn't easy for me to test it thoroughly. I'll work on the issues you helped to find out (thx!).

2. The reason for using .getWeightVariable() is to make the code cleaner. It used to be two if conditions (e.g. evaluating whether options[["weight"]] is equal to ""), so there was quite some repetition of code (especially after adding the three new logistic regression variants). 

3. Passing the vector of values from the dataset would lead to error. The default (no weight is added) is an empty string, which glm does not accept. If a weight variable is used, it has to be added to the glm function as a variable (instead of a string), hence eval(call(get("variable name"))). I tried to simply the implementations but this seems to be the only one that works. 

![image](https://user-images.githubusercontent.com/20680189/199267009-a939ff3a-7fee-476f-8229-1a84b14981b6.png)

Reply:
> The reason for using .getWeightVariable() is to make the code cleaner

Yes that's fine, I was just wondering about the implementation.

> Passing the vector of values from the dataset would lead to error. The default (no weight is added) is an empty string, which glm does not accept. If a weight variable is used, it has to be added to the glm function as a variable (instead of a string), hence eval(call(get("variable name"))). 

Yes, but how about something like

```r
weights <- if(options[["weights"]] == "") NULL else dataset[[options[["weights"]]]] 
...
nullModel <- stats::glm(..., weights = weights)
```
That wouldn't work? :)

> I haven't been able to load the module into JASP (see screenshot for error, any idea what happened?), so it wasn't easy for me to test it thoroughly. 

Hm apart from the usual "check whether you use your personal GitHub PAT" or "click `Clear installed modules and packages` and try again",  or "which JASP version are you using?" not sure, but if nothing helps perhaps it's something related to https://github.com/jasp-stats/INTERNAL-jasp/issues/2151 and needs to be fixed

Reply:
> weights <- if(options[["weights"]] == "") NULL else dataset[[options[["weights"]]]] 
...
nullModel <- stats::glm(..., weights = weights)

Unfortunately, this didn't work. When `weights` isn't `NULL`, the `glm` function would somehow look for a variable called `weights` and of course, this variable doesn't exist. Tried some variations, same problem. 

I think this probably has to do with how R is run behind JASP, because in a normal R session, I can use your suggested approach perfectly fine.

Reply:
hmm okay, i see the issue.

Seems it it due to the following lines in `stats::glm`

```r
mf <- match.call(expand.dots = FALSE)
m <- match(c("formula", "data", "subset", "weights", "na.action", 
             "etastart", "mustart", "offset"), names(mf), 0L)
mf <- mf[c(1L, m)]
mf$drop.unused.levels <- TRUE
mf[[1L]] <- quote(stats::model.frame)
mf <- eval(mf, parent.frame())
```
Where the `model.frame()` fails to find the appropriate object in the `parent.frame()`.
What is kind of weird that `stats::lm` does the same thing, but supplying `weights` as a numeric vector works in our linear regression without issues: https://github.com/jasp-stats/jaspRegression/blob/fcb6de3fd15a80a0e9f7930d3f57c3099bd1c62f/R/regressionlinear.R#L842

@vandenman do you have an idea what could be going wrong here? I guess the workaround by @fqixiang works fine, but it irks me that it does not just work.

Reply:
I'm not sure. These `match.call` constructions that end with `eval(mf, parent.frame())` seem a bit unreliable to me. BAS has a similar problem where an object cannot be found, but only when you pass the formula as an object and not as a literal (see https://github.com/merliseclyde/BAS/issues/56).

Reply:
Just fixed the issue in BAS (finally!!!) and the problem/solution is discussed nicely in https://stackoverflow.com/questions/61164404/call-to-weight-in-lm-within-function-doesnt-evaluate-properly/61164660#61164660?newreg=04a2b71c6da04693a5b172a54a4a43b0

The problem is that `BAS`, `lm` and `glm`assume that the weights are in the same environment as the formula

here is a simple fix with `lm` that is now in `BAS` on GitHub (not CRAN)

`test_that("env and lm", {
  data(UScrime, package = "MASS")
  UScrime <- UScrime[, 1:5]
  
  mylm = function(object) {
    modelform = as.formula(eval(object$call$formula, parent.frame()))
    environment(modelform) = environment()
    data = eval(object$call$data)
    weights = eval(object$call$weights)
    
    object = lm(formula = modelform,
                data = data,
                weights = weights)
   return(object) }
   
  crime.lm1 <- lm(formula = M ~ So + Ed + Po1 + Po2, data = UScrime)  
  tmp1 = mylm(crime.lm1)
  
  form = M ~ So + Ed + Po1 + Po2
  crime.lm2 <- lm(formula = form, data = UScrime)  
  
  tmp = mylm(crime.lm2)
  
  expect_equal(coef(tmp), coef(tmp1))
  
})`